### ISSUE TITLE
Kitchen tests failing on creating /etc/sudoers.d/kitchen

### ISSUE THREAD

**Issue by lifesoccult** [test-kitchen/kitchen-docker#99]
Greetings, I am getting the following error when running kitchen tests with the latest version of kitchen-docker (1.7.0):

`Step 8 : RUN echo 'kitchen ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers.d/kitchen
 ---> Running in 6231397dcf8f
/bin/sh: /etc/sudoers.d/kitchen: No such file or directory`

On another box running version 1.5.0 of kitchen-docker, this problem does not exist as the `RUN echo '#{username} ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers.d/#{username}` is not in the docker.rb file.

As a workaround, I've deleted this line from my docker.rb file. This is on a CentOS 5 docker container. I have not tested against any other docker OS containers.

Thanks,

Jonathan

**Comment by portertech** [comment]
We could add a test for the sudoers directory prior to running that line.

**Comment by lifesoccult** [comment]
Works for me. However, I don't fully understand the need for writing to this file if you're writing to /etc/sudoers already. Could you elaborate on that?

**Comment by miketheman** [comment]
I'm getting the same, on ubuntu 10.04:

```console
      Step 7 : RUN echo kitchen:kitchen | chpasswd
        ---> Running in 77abbcfba4ef
        ---> 635dd16f6e5f
       Removing intermediate container 77abbcfba4ef
       Step 8 : RUN echo 'kitchen ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
        ---> Running in bc6b2bbe6626
        ---> 5b84667c8f50
       Removing intermediate container bc6b2bbe6626
       Step 9 : RUN echo 'kitchen ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers.d/kitchen
        ---> Running in 27eb705f5b7b
        ---> 0e9a05c25706
       Removing intermediate container 27eb705f5b7b
       Successfully built 0e9a05c25706
       e2d225bff003657e050ccda26da669bf2ebe0e7a463094603a57d15fcb4e95b2
       [{
...
       sudo: /etc/sudoers.d/kitchen is mode 0644, should be 0440
       >>> /etc/sudoers.d/README: /etc/sudoers.d/kitchen near line 18 <<<
       sudo: parse error in /etc/sudoers.d/README near line 18

       sudo: no valid sudoers sources found, quitting
```

Looks to have been introduced here: https://github.com/portertech/kitchen-docker/commit/03b2a30d66a7a35ad9f2c5b16511651de6390deb

From the README inside the sudoers.d directory on Ubuntu 12.04:

> As of Debian version 1.7.2p1-1, the default /etc/sudoers file created on
installation of the package now includes the directive:
 	#includedir /etc/sudoers.d
...
Note that there must be at least one file in the sudoers.d directory (this
one will do), and all files in this directory should be mode 0440.

I believe that older CentOS versions didn't have the `sudoers.d` logic at all.

I can confirm that downgrading to 1.5.0 solves this particular problem.

Out of curiosity, why was the sudoers.d method introduced?

**Comment by miketheman** [comment]
Also confirmed that /etc/sudoers.d does not exist in Cent5, and does in Cent6.

**Comment by brandocorp** [comment]
I can confirm I'm currently seeing this on ubuntu 12.04

```
-----> Creating <default-ubuntu-1204>...
       Sending build context to Docker daemon  2.56 kB
       Sending build context to Docker daemon 
       Step 0 : FROM ubuntu:12.04
        ---> f959d044ebdf
       Step 1 : RUN dpkg-divert --local --rename --add /sbin/initctl
        ---> Running in f81f842d9775
       Leaving 'local diversion of /sbin/initctl to /sbin/initctl.distrib'
        ---> 3c2f332f2047
       Removing intermediate container f81f842d9775
       Step 2 : RUN ln -sf /bin/true /sbin/initctl
        ---> Running in 3b40f99a4021
        ---> aab0df78640c
       Removing intermediate container 3b40f99a4021
       Step 3 : ENV DEBIAN_FRONTEND noninteractive
        ---> Running in 3679ec52a771
        ---> 6204a11c5de1
       Removing intermediate container 3679ec52a771
       Step 4 : RUN apt-get update
        ---> Running in a51fff60e496
       Hit http://archive.ubuntu.com precise Release.gpg
        ...
       Get:31 http://archive.ubuntu.com precise-security/universe i386 Packages [145 kB]
       Fetched 28.9 MB in 44s (654 kB/s)
       Reading package lists...
        ---> c0219b03987c
       Removing intermediate container a51fff60e496
       Step 5 : RUN apt-get install -y sudo openssh-server curl lsb-release
        ---> Running in 68e62dcfff57
       Reading package lists...
       Building dependency tree...
       Reading state information...
       The following extra packages will be installed:
        ...
       Suggested packages:
        ...
       The following NEW packages will be installed:
        ...
       The following packages will be upgraded:
         python2.7-minimal
       1 upgraded, 54 newly installed, 0 to remove and 9 not upgraded.
       Need to get 12.4 MB of archives.
       After this operation, 34.9 MB of additional disk space will be used.
       Get:1 http://archive.ubuntu.com/ubuntu/ precise-updates/main libsqlite3-0 amd64 3.7.9-2ubuntu1.1 [349 kB]
        ...
       Get:55 http://archive.ubuntu.com/ubuntu/ precise/main ssh-import-id all 2.10-0ubuntu1 [6598 B]
       debconf: delaying package configuration, since apt-utils is not installed
       Fetched 12.4 MB in 14s (862 kB/s)
        ...
       Processing triggers for libc-bin ...
       ldconfig deferred processing now taking place
        ---> e2523991a2e1
       Removing intermediate container 68e62dcfff57
       Step 6 : RUN useradd -d /home/kitchen -m -s /bin/bash kitchen
        ---> Running in 423a5001b6cf
        ---> 44f551f6494b
       Removing intermediate container 423a5001b6cf
       Step 7 : RUN echo kitchen:kitchen | chpasswd
        ---> Running in b2bb47768da3
        ---> 5d927be264b1
       Removing intermediate container b2bb47768da3
       Step 8 : RUN echo 'kitchen ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
        ---> Running in 2a1e6a8146a2
        ---> 63dc8108e171
       Removing intermediate container 2a1e6a8146a2
       Step 9 : RUN echo 'kitchen ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers.d/kitchen
        ---> Running in 885a960552df
        ---> 10b518f69700
       Removing intermediate container 885a960552df
       Step 10 : RUN curl -L https://www.opscode.com/chef/install.sh | bash
        ---> Running in 89816ff9d43e
         % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                        Dload  Upload   Total   Spent    Left  Speed
100 18378  100 18378    0     0  39025      0 --:--:-- --:--:-- --:--:-- 57431
       Downloading Chef  for ubuntu...
       downloading https://www.opscode.com/chef/metadata?v=&prerelease=false&nightlies=false&p=ubuntu&pv=12.04&m=x86_64
         to file /tmp/install.sh.9/metadata.txt
       trying wget...
       url	https://opscode-omnibus-packages.s3.amazonaws.com/ubuntu/12.04/x86_64/chef_12.0.3-1_amd64.deb
       md5	e849cf61d3e3fd37224c11f9902aa536
       sha256	a8e9c78cc9b9a17a32b9b4ae1650b8b6454b4f0abd611048d099a659ca5de28f
       downloaded metadata file looks valid...
       downloading https://opscode-omnibus-packages.s3.amazonaws.com/ubuntu/12.04/x86_64/chef_12.0.3-1_amd64.deb
         to file /tmp/install.sh.9/chef_12.0.3-1_amd64.deb
       trying wget...
       Comparing checksum with sha256sum...
       Installing Chef 
       installing with dpkg...
       Selecting previously unselected package chef.
       (Reading database ... 9171 files and directories currently installed.)
       Unpacking chef (from .../chef_12.0.3-1_amd64.deb) ...
       Setting up chef (12.0.3-1) ...
       Thank you for installing Chef!
        ---> dc77ed109c39
       Removing intermediate container 89816ff9d43e
       Successfully built dc77ed109c39
       Warning: '-privileged' is deprecated, it will be replaced by '--privileged' soon. See usage.
       cd7c225ebb1af3299bc3291afc68b431c8ed8660159326c061d84cea9ab081e9
       [{
           "AppArmorProfile": "",
           "Args": [
        "-D",
        "-o",
        "UseDNS=no",
        "-o",
        "UsePAM=no",
        "-o",
        "PasswordAuthentication=yes",
        "-o",
        "UsePrivilegeSeparation=no",
        "-o",
        "PidFile=/tmp/sshd.pid"
           ],
           "Config": {
        "AttachStderr": false,
        "AttachStdin": false,
        "AttachStdout": false,
        "Cmd": [
            "/usr/sbin/sshd",
            "-D",
            "-o",
            "UseDNS=no",
            "-o",
            "UsePAM=no",
            "-o",
            "PasswordAuthentication=yes",
            "-o",
            "UsePrivilegeSeparation=no",
            "-o",
            "PidFile=/tmp/sshd.pid"
        ],
        "CpuShares": 0,
        "Cpuset": "",
        "Domainname": "",
        "Entrypoint": null,
        "Env": [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            "DEBIAN_FRONTEND=noninteractive"
        ],
        "ExposedPorts": {
            "22/tcp": {}
        },
        "Hostname": "cd7c225ebb1a",
        "Image": "dc77ed109c39",
        "MacAddress": "",
        "Memory": 0,
        "MemorySwap": 0,
        "NetworkDisabled": false,
        "OnBuild": null,
        "OpenStdin": false,
        "PortSpecs": null,
        "StdinOnce": false,
        "Tty": false,
        "User": "",
        "Volumes": null,
        "WorkingDir": ""
           },
           "Created": "2015-02-10T23:50:35.881680058Z",
           "Driver": "aufs",
           "ExecDriver": "native-0.2",
           "HostConfig": {
        "Binds": null,
        "CapAdd": null,
        "CapDrop": null,
        "ContainerIDFile": "",
        "Devices": [],
        "Dns": null,
        "DnsSearch": null,
        "ExtraHosts": null,
        "IpcMode": "",
        "Links": null,
        "LxcConf": [],
        "NetworkMode": "bridge",
        "PortBindings": {
            "22/tcp": [
                {
                    "HostIp": "",
                    "HostPort": ""
                }
            ]
        },
        "Privileged": true,
        "PublishAllPorts": false,
        "RestartPolicy": {
            "MaximumRetryCount": 0,
            "Name": ""
        },
        "SecurityOpt": null,
        "VolumesFrom": null
           },
           "HostnamePath": "/mnt/sda1/var/lib/docker/containers/cd7c225ebb1af3299bc3291afc68b431c8ed8660159326c061d84cea9ab081e9/hostname",
           "HostsPath": "/mnt/sda1/var/lib/docker/containers/cd7c225ebb1af3299bc3291afc68b431c8ed8660159326c061d84cea9ab081e9/hosts",
           "Id": "cd7c225ebb1af3299bc3291afc68b431c8ed8660159326c061d84cea9ab081e9",
           "Image": "dc77ed109c39a8156c9a02a67ca69c467abae0049fb4411df23fe0df4d9c4f69",
           "MountLabel": "",
           "Name": "/tender_pike",
           "NetworkSettings": {
        "Bridge": "docker0",
        "Gateway": "172.17.42.1",
        "IPAddress": "172.17.0.110",
        "IPPrefixLen": 16,
        "MacAddress": "02:42:ac:11:00:6e",
        "PortMapping": null,
        "Ports": {
            "22/tcp": [
                {
                    "HostIp": "0.0.0.0",
                    "HostPort": "49165"
                }
            ]
        }
           },
           "Path": "/usr/sbin/sshd",
           "ProcessLabel": "",
           "ResolvConfPath": "/mnt/sda1/var/lib/docker/containers/cd7c225ebb1af3299bc3291afc68b431c8ed8660159326c061d84cea9ab081e9/resolv.conf",
           "State": {
        "Error": "",
        "ExitCode": 0,
        "FinishedAt": "0001-01-01T00:00:00Z",
        "OOMKilled": false,
        "Paused": false,
        "Pid": 8879,
        "Restarting": false,
        "Running": true,
        "StartedAt": "2015-02-10T23:50:36.239977741Z"
           },
           "Volumes": {},
           "VolumesRW": {}
       }
       ]
       Finished creating <default-ubuntu-1204> (2m25.75s).
-----> Converging <default-ubuntu-1204>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 3.2.3...
       Removing non-cookbook files before transfer
       Preparing data_bags
       Preparing environments
       Preparing validation.pem
       Preparing client.rb
       sudo: /etc/sudoers.d/kitchen is mode 0644, should be 0440
       Transferring files to <default-ubuntu-1204>
       sudo: /etc/sudoers.d/kitchen is mode 0644, should be 0440
       Starting Chef Client, version 12.0.3
       Creating a new client identity for default-ubuntu-1204 using the validator key.
        ...
       
       
       Running handlers:
       Running handlers complete
       Chef Client finished, 26/28 resources updated in 8.338013181 seconds
       Finished converging <default-ubuntu-1204> (0m19.53s).
```

**Comment by spheromak** [comment]
also seeing issues running centos5 containers. Centos5 doesn't have /etc/sudoers.d nor does the version of sudo on cent5 support it. 

:+1:  for testing first, but also note that some platforms don't support includes.

**Comment by franklinwise** [comment]
seeing the same as @brandocorp on ubuntu-12.04

sudo: /etc/sudoers.d/kitchen is mode 0644, should be 0440

**Comment by CpuID** [comment]
This should be resolved for centos5, can't speak for ubuntu-12.04

**Comment by jeffreycoe** [comment]
Closing issue due to inactivity. Please open a new issue if there is still a problem. Thanks!
